core/watcher/eventsource/notify.go

```diff
@@ -0,0 +1,148 @@
+// channel, once the mapper has processed them. Filtering of values is done
+// first by the filter, and then by the mapper.
+// A filter option is required, though additional filter options can be
+// provided.
+func NewNotifyMapperWatcher(
+	base *BaseWatcher, mapper Mapper,
+	filterOption FilterOption, filterOptions ...FilterOption,
+) (*NotifyWatcher, error) {
+	filters := append([]FilterOption{filterOption}, filterOptions...)
+	opts := make([]changestream.SubscriptionOption, len(filters))
+	for i, opt := range filters {
+		predicate := opt.ChangePredicate()
+		if predicate == nil {
+			return nil, errors.Errorf("no change predicate provided for filter option %d", i)
+		}
+
+		opts[i] = changestream.FilteredNamespace(opt.Namespace(), opt.ChangeMask(), func(e changestream.ChangeEvent) bool {
+			return predicate(e.Changed())
+		})
+	}
+
+	w := &NotifyWatcher{
+		BaseWatcher: base,
+		out:         make(chan struct{}),
+		filterOpts:  opts,
+		mapper:      mapper,
+	}
+
+	w.tomb.Go(w.loop)
+	return w, nil
+}
+
+// Changes will emit a struct{} on the channel, which will be the coalesced set
+// of changes within a given term.
+func (w *NotifyWatcher) Changes() <-chan struct{} {
+	return w.out
+}
+
+func (w *NotifyWatcher) loop() error {
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	defer close(w.out)
+
+	subscription, err := w.watchableDB.Subscribe(w.filterOpts...)
+	if err != nil {
+		return errors.Annotatef(err, "subscribing to namespaces")
+	}
+	defer subscription.Unsubscribe()
+
+	// By reassigning the in and out channels, we effectively ticktock between
+	// read mode and dispatch mode. This ensures we always dispatch
+	// notifications for changes we received before reading more, and every
+	// channel read/write is guarded by checks of the tomb and subscription
+	// liveness.
+	// We begin in dispatch mode in order to send the initial notification.
+	in := subscription.Changes()
+	out := w.out
+
+	w.drainInitialEvent(ctx, in)
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-subscription.Done():
+			return ErrSubscriptionClosed
+		case changes, ok := <-in:
+			if !ok {
+				w.logger.Debugf(ctx, "change channel closed; terminating watcher")
+				return nil
+			}
+
+			// Allow the possibility of the mapper to drop/filter events.
+			changed, err := w.mapper(ctx, w.watchableDB, changes)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			// If the mapper has dropped all events, we don't need to do
+			// anything.
+			if len(changed) == 0 {
+				continue
+			}
+
+			// We have changes. Tick over to dispatch mode.
+			out = w.out
+		case out <- struct{}{}:
+			// We have dispatched. Tick over to read mode.
+			out = nil
+		}
+	}
+}
+
+func (w *NotifyWatcher) drainInitialEvent(ctx context.Context, in <-chan []changestream.ChangeEvent) {
+	select {
+	case _, ok := <-in:
+		if !ok {
+			w.logger.Debugf(ctx, "change channel closed; terminating watcher")
+			return
+		}
+	default:
```

Comment:
Is this just a "maybe drain initial event"? Bit of a smell.

Reply:
There is a chance that the fillter options for the subscription are too wide for a single notification. So what can happen is, that you'll receive multiple changes for the initial subscription; essentially you get a double pump. Ultimately there is no harm in this, it does cause more load on the system creating potentially additional churn.

There is also an additional benefit, by coalescing at this level, we can reduce the dependency on sequencing of notifications.